### PR TITLE
Add the internal experimental metric feature package

### PR DESCRIPTION
**Pull request by MrAlias**
Used to enable specification defined features. Such as:

- https://github.com/open-telemetry/opentelemetry-go/pull/4457
- https://github.com/open-telemetry/opentelemetry-go/pull/4455

This is the first step in adding these features.

### CODE REVIEW

**Comment by MrAlias** [comment]
> Did you consider modeling this after feature-gates in the collector? https://github.com/open-telemetry/opentelemetry-collector/tree/main/featuregate. If we used it directly, it would mean collector users could configure OTel Go's feature gates using the collector's feature-gates flag, which would be neat. We would have to wait for the v1 release to avoid a circular unstable dependency.
> 
> Assuming we don't want to depend on it directly, could we follow the same pattern, and use a OTEL_GO_FEATURE_GATES env var with k0=v0,k1=v1 format, and defined stages? Or is that overkill?

I haven't looked at this yet. I'll be sure to take a look.

One question I have though, is how will uses distinguish between their SDK features and collector features? For example, given we will support different features than the collector but a user may define them in the same environment variable does that mean we need to drop unknown? Do we log this error?

**Comment by dashpole** [comment]
> One question I have though, is how will uses distinguish between their SDK features and collector features? For example, given we will support different features than the collector but a user may define them in the same environment variable does that mean we need to drop unknown? Do we log this error?

We would have to decide how to handle errors from parsing the flags, including when users set feature gates that don't exist. Writing to the global error handler seems reasonable.

**Comment by MrAlias** [comment]
@dashpole I don't see in that linked package where the environment variable `OTEL_GO_FEATURE_GATES` is supported. Is this an extension somewhere, or am I just bad at searching?

**Comment by MrAlias** [comment]
@dashpole does `featuregate` support processing configuration? That was an ask from [here](https://github.com/open-telemetry/opentelemetry-go/pull/4457#discussion_r1341454330) that I think is useful.

We would need to support additional environment variables if it does right? Or could we add it to that package?

**Comment by dashpole** [comment]
It does not support configuration.  We would need additional env vars or something else.

**Comment by dashpole** [comment]
> I don't see in that linked package where the environment variable OTEL_GO_FEATURE_GATES is supported. Is this an extension somewhere, or am I just bad at searching?

We would have to add that.  It only supports flags today.  Or, we could require users to register flags to use it, which has the advantage of making users decide how to handle parsing failures (so they don't need to use the global error handler)

**Review by dashpole — APPROVED** [review]

File: sdk/metric/internal/x/x.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+// Key returns the environment variable key that needs to be set to enable the
+// feature.
+func (f Feature[T]) Key() string { return f.key }
+
+// Lookup returns the user configured value for the feature and true if the
+// user has enabled the feature. Otherwise, if the feature is not enabled, a
+// zero-value and false are returned.
+func (f Feature[T]) Lookup() (v T, ok bool) {
+	vRaw, present := os.LookupEnv(f.key)
+	if !present {
```
Comment (pellared) [inline sdk/metric/internal/x/x.go:79]: > The SDK MUST interpret an empty value of an environment variable the same way as when the variable is unset.

From https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md#parsing-empty-value

```suggestion
	vRaw, present := os.GetEnv(f.key)
	if vRaw == "" {
```

**Review by pellared — COMMENTED** [review]
Can we add tests?

File: sdk/metric/internal/x/x.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	//
+	// To enable this feature set the OTEL_GO_X_EXEMPLAR environment variable
+	// to the case-insensitive string value of "true" (i.e. "True" and "TRUE"
+	// will also enable this).
+	Exemplars = newFeature("EXEMPLAR", func(v string) (string, bool) {
+		if strings.ToLower(v) == "true" {
+			return v, true
+		}
+		return "", false
+	})
```
Comment (pellared) [inline sdk/metric/internal/x/x.go:39]: nit: Why are we returning `(string, bool)` and not e.g. `(struct{}, bool)`. Any reason why to return the value?

**Review by pellared — APPROVED** [review]

File: sdk/metric/internal/x/x_test.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	}
+}
+
+func setenv(k, v string) func(t *testing.T) {
+	return func(t *testing.T) { t.Setenv(k, v) }
+}
+
+func evalEnabled[T any](f Feature[T], want T) func(*testing.T) {
+	return func(t *testing.T) {
+		t.Helper()
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/metric/internal/x/x_test.go:58]: `t.Helper()` ensures the function is included in the error trace:

```console
--- FAIL: TestExemplars (0.00s)
    --- FAIL: TestExemplars/true (0.00s)
        x_test.go:28:
            	Error Trace:	/home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/x/x_test.go:63
            	            				/home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/x/x_test.go:47
            	Error:      	Not equal:
            	            	expected: "truE"
            	            	actual  : "true"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1 +1 @@
            	            	-truE
            	            	+true
            	Test:       	TestExemplars/true
            	Messages:   	Lookup value
FAIL
FAIL	go.opentelemetry.io/otel/sdk/metric/internal/x	0.003s
FAIL
```

Having it ensures the test doesn't just say `/home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/x/x_test.go:47`.

File: sdk/metric/internal/x/x_test.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+		assert.False(t, f.Enabled(), "enabled")
+
+		v, ok := f.Lookup()
+		assert.False(t, ok, "Lookup state")
+		assert.Equal(t, zero, v, "Lookup value")
+	}
+}
```
[the remark this answers is not in the document]
Comment (pellared) [reply sdk/metric/internal/x/x_test.go]: you can add

```go
func testUnset[T any](t *testing.Tf Feature[T]) {
	t.Run(f.Key()+" unset", func(t *testing.T) {
		assert.False(t, f.Enabled(), "enabled")

		var zero T
		got, ok := f.Lookup()
		assert.False(t, ok, "Lookup state")
		assert.Equal(t, zero, got, "Lookup value")
	})
}
```

File: sdk/metric/internal/x/x_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pellared) [reply sdk/metric/internal/x/x_test.go:61]: Wow. This is strange

> Helper marks the calling function as a test helper function. When printing file and line information, that function will be skipped.

File: sdk/metric/internal/x/x_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (pellared) [reply sdk/metric/internal/x/x_test.go]: If you insist in the current pattern (which is not how most tests look like and does not look familiar for me) then I would at least rename `eval[Enabled|Disabled]` to `assert[Enabled|Disabled]` or `verify[Enabled|Disabled]`  to make it more clear that these are veryfing steps.

**Review by pellared — APPROVED** [review]

**Pull request merged**